scripts/build_climate_projections.py

```diff
@@ -0,0 +1,210 @@
+"""
+
+
+import datetime as dt
+import os
+
+import atlite
+import geopandas as gpd
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pypsa
+import xarray as xr
+from _helpers import configure_logging, create_logger
+from shapely.geometry import LineString as Line
+from shapely.geometry import Point
+
+logger = create_logger(__name__)
+
+
+def crop_cmip6(cmip6_xr, cutout_xr):
+    # spartial margin needed to avoid data losses during further interpolation
+    d_pad = 1
+
+    cmip6_region = cmip6_xr.sel(
+        lat=slice(
+            min(cutout_xr.coords["y"].values) - d_pad,
+            max(cutout_xr.coords["y"].values) + d_pad,
+        ),
+        lon=slice(
+            min(cutout_xr.coords["x"].values) - d_pad,
+            max(cutout_xr.coords["x"].values) + d_pad,
+        ),
+    )
+    return cmip6_region
+
+
+def interpolate_cmip6_to_cutout_grid(cmip6_xr, cutout_xr):
+    # TODO read from the cutout file instead of hardcoding
+    dx_new = 0.3
+
+    newlon = np.arange(
+        round(min(cutout_xr.coords["x"].values), 1),
+        round(max(cutout_xr.coords["x"].values) + dx_new, 1),
+        dx_new,
+    )
+    newlat = np.arange(
+        round(min(cutout_xr.coords["y"].values), 1),
+        round(max(cutout_xr.coords["y"].values) + dx_new, 1),
+        dx_new,
+    )
+    cmip6_interp = cmip6_xr.interp(time=cmip6_xr.time, lat=newlat, lon=newlon)
+
+    return cmip6_interp
+
+
+# TODO fix years_window
+def subset_by_time(cmip6_xr, month, year, years_window):
+    # TODO add warning in case there are no enough years
+    cmip6_in_period = cmip6_xr.where(
+        (
+            (cmip6_xr["time.month"] == month)
+            & (cmip6_xr["time.year"] >= year - years_window)
+            & (cmip6_xr["time.year"] < year + years_window - 1)
+        ),
+        drop=True,
+    )
+    return cmip6_in_period
+
+
+def calculate_proj_of_average(cmip6_xr, month, year0, year1, years_window):
+    cmip6_interp_year0 = subset_by_time(
+        cmip6_xr,
+        month,
+        year=year0,
+    )
+    cmip6_interp_year1 = subset_by_time(cmip6_xr, month=month, year=year1)
+    dt_interp = cmip6_interp_year1["t"].mean("member").mean(
+        "time"
+    ) - cmip6_interp_year0["t"].mean("member").mean("time")
+    return dt_interp
+
+
+def build_projection_for_month(cutout_xr, dt_xr, month):
+    k_time = cutout_xr.time.dt.month.isin(month)
+
+    for i in range(0, len(cutout_xr.y)):
+        for j in range(0, len(cutout_xr.x)):
+            cutout_xr.temperature[k_time, i, j] = np.add(
+                cutout_xr.temperature[k_time, i, j],
+                np.array([dt_xr[i, j].item()] * k_time.sum().item()),
+            )
```

Comment:
At first sight, it feels like this code may take long time to compute.
What are you trying to do here?

Reply:
I need to modify a temperature value at each spatial point according to a projection for an increase in the monthly temperature. You are absolutely right that it may be quite time-consuming. Do you have any ideas on possible approaches to increase performance?